data-processing/entities/sentences/extractor.py

```diff
@@ -45,22 +45,27 @@ def parse(self, tex_path: str, tex: str) -> Iterator[Sentence]:
 
         if last_segment is not None:
             plaintext_to_tex_offset_map[len(plaintext)] = last_segment.tex_end
-
         # Segment the plaintext. Return offsets for each setence relative to the TeX input
         segmenter = pysbd.Segmenter(language="en", clean=False, char_span=True)
+        # Record the current length of the plain text so account for the extractor bug
+        length_so_far_in_plain_text = 0
         for i, sentence in enumerate(segmenter.segment(plaintext)):
+            # Since the sentence extractor has several bugs related to finding start and end indicies
+            # we will simply set them ourselves.
+            real_start = plaintext.find(sentence.sent, length_so_far_in_plain_text)
```

Comment:
This looks great! A couple of questions to make this code super robust:

1. Can you add a check for the return value `-1` (when the sentence is not found) with a `logging.warning` statement? I don't expect this to happen, but over millions of papers it may come up.

2. Can you check to see if `pysbd` ever changes the text in the `sentence.sent` from how the text is written in the input text (i.e. does it remove invalid characters from `sent`?) If the sentences will always be a substring of the plaintext, this looks just fine! If the sentences are sometimes transformed, then `find` will definitely return `-1` sometimes. Then we should add another rule that does something useful if `find` returns `-1`), for example, sets `real_start` at an offset of one character after the last sentence's `real_end`.

Reply:
Gotcha :). 
1. I know for a fact after looking through the actual pysbd that it will indeed remove some very very obscure special characters (these are used in the actual pysbd implementation). Maybe I can combine your point number 1 and 2 and just have a logging.warning statement that says "Warning failed to parse sentence, skipping sentence"?
2. I also wanted to reiterate that psybd fails for sentences with more than one punctuation that are not periods (say !! or !?). I'm honestly not really sure how to handle this, throwing the psybd parser in a try catch sounds like it could be very expensive over thousands of iterations, but I'm not sure if there is a better way? Maybe just write a regex expr and just don't search the string if it's true?

Reply:
> throwing the psybd parser in a try catch sounds like it could be very expensive over thousands of iterations, but I'm not sure if there is a better way?

This would probably be fine, as I imagine the exceptions would be pretty rare, and string searching is one of the fastest parts of the pipeline currently.

> Maybe just write a regex expr and just don't search the string if it's true?

I think this could work! What type of regex do you have in mind?

Reply:
Gotcha, I'll just the try catch then bc it's going to be more robust then a regex

Reply:
Actually I guess there really isn't a point to a try catch now that I think about it? We would probably want it to fail as otherwise the rest of the pipeline wont work and it will be mislead

Reply:
@zkirby I think the hope is we can keep extracting as many other sentences as we can even if there's a failure for one sentence. So I think the ideal behavior is:

* flag if the positions couldn't be recovered for this sentence
* keep going, recovering the correct position of the sentences after this one if we can

Reply:
There are tiny exceptions like this everywhere in the pipeline, so the typical approach has been to try to catch unexpected errors, log them, and keep processing as much of the paper as possible. This tends to work pretty well in papers where there are hundreds of things that need to be processed.

Reply:
Gotcha! I think the double punctuation is actually another indexing error unfortunately which would mean using a try/catch for every sentence. Not sure if this is okay?

Reply:
Probably better to open another PR and try to fix it?

Reply:
@zkirby does the double punctuation cause an exception, or just split sentences in a way that weren't expecting?

Reply:
I'm game with rolling it into this PR, I'm just realizing I didn't catch your meaning about the double punctuation initially.

Reply:
It throws an index out of bounds in plaintext_to_tex_offset_map. Any ideas?

Reply:
Just getting a bit more context (and also sorry if you've already answered this XD), why does it cause an index out of bounds exception? Can you give me a brief code example?

Reply:
Ah, as far as why I'm not sure I can spend sometime investigating right now if we are going to include it in the PR, but for example a sentence like 'hello. world!!!' returns: 
```
File "/data-processing/entities/sentences/extractor.py", line 54, in parse
    end = plaintext_to_tex_offset_map[sentence.end]
KeyError: 21
```